exp/operator/controllers/shared.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cluster-api/api/v1alpha4"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func preflightChecks(ctx context.Context, c client.Client, gvk schema.GroupVersionKind) (*v1alpha4.Condition, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(4).Info("Performing flight checks.")
+
+	providerList := &unstructured.UnstructuredList{}
+	providerList.SetGroupVersionKind(gvk)
+
+	if err := c.List(ctx, providerList); err != nil {
+		return nil, errors.Wrap(err, "failed to list providers")
+	}
+
+	// Checking if more than one instance of the same provider exists in the cluster.
+	if len(providerList.Items) > 1 {
```

Comment:
Sorry to go back to this point but I don't really grasp `Len()>1`

If there is already more than one provider of a given kind/name that's means that the cluster is already in a not supported state, so we are defeting the entire scope of preflight (prevents to get to a not supported state).

Also, we really need to check by kind/name, because it is perfectly acceptable to have two infrastructure providers (AWS and VSphere) or two bootstrap providers (kubeadm and Talos). What we should prevent instead is to have two instances of the same provider (e.g. two infrastructure AWS)

Reply:
Ok, I get it now. Thanks. We can get a list of all providers and check if there is no duplicate name. I think resource name is the only type indicator since it's also going to be used for fetching configuration.

Reply:
Do we not still want to have validation that there is only one `CoreProvider`, I get what is being said here for other kinds of providers but IIUC the `CoreProvider` is a singleton and cannot be duplicated? Maybe this is a case of the logic needing to be moved elsewhere?

Reply:
I updated the logic and added infrastructure providers to unit tests, PTAL